openhands/server/session/manager.py

```diff
@@ -1,25 +1,29 @@
+import asyncio
 import time
-from dataclasses import dataclass
+from dataclasses import dataclass, field
 
 from fastapi import WebSocket
+import socketio
 
 from openhands.core.config import AppConfig
 from openhands.core.logger import openhands_logger as logger
+from openhands.core.schema.agent import AgentState
+from openhands.events.event import EventSource
+from openhands.events.observation.agent import AgentStateChangedObservation
+from openhands.events.serialization.event import event_to_dict
 from openhands.events.stream import session_exists
 from openhands.server.session.conversation import Conversation
 from openhands.server.session.session import Session
 from openhands.storage.files import FileStore
+from openhands.utils.shutdown_listener import should_continue
 
 
 @dataclass
 class SessionManager:
     config: AppConfig
     file_store: FileStore
-
-    def add_or_restart_session(self, sid: str, ws_conn: WebSocket) -> Session:
-        return Session(
-            sid=sid, file_store=self.file_store, ws=ws_conn, config=self.config
-        )
+    local_sessions_by_sid: dict[str, Session] = field(default_factory=dict)
+    local_sessions_by_connection_id: dict[str, Session] = field(default_factory=dict)
```

Comment:
do we really need to track both? seems dangerous...

Reply:
In listen.py - there is a disconnect method that is called within a reasonable time when a connection terminates. This calls `await session_manager.disconnect_from_local_session` - and this removes the connection id for that session - when all connection ids are removed, a timer starts that stops the runtime 15 seconds after the last disconnect.
In my testing, this means there is no accumulation of connection_ids or session_ids. (This is also why I have to track both)